### PR TITLE
Modifications to support Python 3 and wxPython Phoenix (4.0) - WIP

### PR DESCRIPTION
**Pull request by SamuelDudley**
I have opened this WIP to get an idea for what would be required to support wxPython Phoenix and python 3.x. This WIP is a quick hack and contains breaking changes for 2.x Python versions. Tested with Python 3.5 and 4.0.0a3 release of wxPython Phoenix.

- Windows has pickling errors with the map module (unsure how badly it breaks the map functionality), no issues under linux
- Currently mission editor does not run due to annoying pickling errors (windows and linux)
- PyMAVLink is OK under Python3.x, but some of the tools (like DF log reader) do not work
- Pyserial, Matplotlib, Numpy and OpenCV 3.2 (built from source but a pip package exists) are all OK  under Python3.x
- Not sure how these changes will impact OSX, I need a willing tester.
- A few of the changes will be compatible with Python2.7. It might be worth isolating them and applying those changes first

### CODE REVIEW

**Comment by stephendade** [comment]
> Would this be the last thing stopping MAVProxy supporting Python 3.x?

You'd want to double-check that the pymavlink port to Python 3.x has been completed. I'm _fairly_ certain it has though.

All standard libraries other than that should be ok - we'd just have to check they all have Python 3.x packages.

Finally, I suspect they'll be some syntax changes required in MAVProxy itself.

But, WxPython is by far the largest item stopping us from Python 3 compatibility.

Great work :)

**Comment by stephendade** [comment]
You'd also want to ensure that nothing's broken in the Windows build too

**Comment by stephendade** [comment]
Ahh, OK. Outsourced to me in that case :)

**Comment by SamuelDudley** [comment]
Alright! Machete coding aside, that last commit is everything needed to get all the default modules +map +wxconsole running in python 3.5.
screenshot from 2017-06-14 01-02-56

**Comment by stephendade** [comment]
Just quickly tested console and map - they appear to work in Windows. Will do more comprehensive testing later.

**Comment by SamuelDudley** [comment]
Good to hear and thanks for testing :) let me know if you hit any issues. What other GUI elements are currently in the main build / supported? 
- Mission editor
- Mag cal GUI
- All the cuav modules 
- Mavexplorer
Am I missing anything else?

**Comment by stephendade** [comment]
There's some common stuff like File Open/Save dialogs and menus in mp_menu.py and some of the other files in there.

Also the graph UI, Horizon module, Checklist module.

I have a sneaking suspicion that there'll be some Python2->3 syntax issues in some areas, so you'll be wanting to test all the modules for Python3 compatibility.

We really need a unit test suite :(

**Comment by SamuelDudley** [comment]
So heres an interesting problem I have found. Multiprocessing is not starting child processes during runtime. I can use a launch command like this `--master tcp:127.0.0.1:5760 --sitl 127.0.0.1:5501 --out 127.0.0.1:14551 --out 127.0.0.1:14550 --console --map  --cmd="module load checklist; module load horizon; module load graph; graph SIMSTATE.roll"` and everything is fine, but attempting to start a module which uses multiprocessing (e.g. map, graph, checklist, horizon) with `module load XXX` will fail silently...
@stephendade try running MAVProxy without the `--map` argument and then type `module load map`. On my system the map child process never starts correctly. Really odd...

screenshot from 2017-06-15 12-50-15

File: MAVProxy/mavproxy.py [hunk 1]
```diff
@@ -847,9 +848,13 @@ def input_loop():
...
             sys.exit(1)
-        mpstate.input_queue.put(line)
-
-
+        mpstate.input_count += 1
+        cmds = line.split(';')
+        if len(cmds) == 1 and cmds[0] == "":
+              mpstate.empty_input_count += 1
+        for c in cmds:
+            process_stdin(c)
```
Comment (SamuelDudley) [inline MAVProxy/mavproxy.py]: calling `module load map` here works, but the same command here https://github.com/SamuelDudley/MAVProxy/blob/357239cb97fd6b50d1fc8087b24502bfc75b67fa/MAVProxy/mavproxy.py#L771 fails silently. 
It only seems to affect multiprocessing, child threads can run fine.

**Comment by stephendade** [comment]
The changes you made since last I tested have broken the Windows build:
capture

When I try running without --console:
capture2

**Comment by SamuelDudley** [comment]
@stephendade Can you try the latest commit when you have a chance please. Windows does not support forkserver,so I have changed the multiprocessing type to spawn

**Comment by stephendade** [comment]
You're quick!

I'm getting errors about urllib2 not being found in the help module (same as before).

Console and map work quite nicely now. MAVProxy completely freezes when I go to the settings dialog and click "cancel".

Get the following error when I click on the map:
capture

Also, nothing happens when a enter "param show RC1_MIN" ... I think the param module is not completely working. "param show all" works though.

**Comment by SamuelDudley** [comment]
Ah thanks. I can fix the params and help. The settings screen is annoying but I can try a few things there. Not sure where to start with the map issue, but I'll have a poke around.

**Comment by SamuelDudley** [comment]
Alright!
-  Param and help modules should be good now. note that pymavlink needs some changes to correctly detect a param has been updated.
-  I have attempted to fix the settings widget. Let me know if that is still broken for you.
-  Lastly I have added some prints to where I think the ImagePanel object is trying to be pickled. Can you please dump the text into a gist or a reply here for me.
Thanks!

**Comment by stephendade** [comment]
More results:
-param module now works
-help module has the following error:
capture2

-Settings widget works
-Imagepanel output here:
capture

File: MAVProxy/modules/mavproxy_help.py [hunk 2]
```diff
@@ -22,31 +23,34 @@ def __init__(self, mpstate):
         try:
             import pkg_resources
             self.version = pkg_resources.require("mavproxy")[0].version
-        except:
+            # what error occurs here in windows? ImportError?
+        except ImportError: # we need to catch the explicit error
```
Comment (SamuelDudley) [inline MAVProxy/modules/mavproxy_help.py]: What error will windows raise here?

**Comment by SamuelDudley** [comment]
Sorry about the help module... I have made some changes but I'm unsure what error windows will raise [here](https://github.com/ArduPilot/MAVProxy/pull/420/commits/18b9d21493746500b880ef272bea011a5716b1c0#diff-d5388a8992817d50ce043615851415d5R27). I need to catch the more specific error as the except statement was causing problems for me. As I'm running from source the two existing ways of determining the version of MAVProxy caused exceptions.

Re the ImagePanel pickle... I'm stumped at this point... Any thoughts would be appreciated!

**Comment by stephendade** [comment]
Here's the help module output:
capture

Param help works :)

For the imagepanel pickle - I remember similar errors happening when I was doing the initial Windows conversion for MAVProxy. It happended when multiprocessing attempted to send a variable from one thread to another. In Linux it would just use shared memory, whereas in Windows it would have to package up the variable (via pickle) and send it across (via a queue I think). So most likely cross-threading shenangians.

**Comment by SamuelDudley** [comment]
That help module error should be fixed for you with the last set of commits. I think your on the right track with the pickling issue... I have spent a few hours attempting to get mission editor to run without success facing a range of cryptic pickling errors :( Unfortunately the stack trace is completely useless...

**Comment by stephendade** [comment]
I now get a different error when loading up the help module:
capture

**Comment by SamuelDudley** [comment]
Thanks for testing... Not sure why I wasn't having that issue? I'll have a look at it :)

**Comment by SamuelDudley** [comment]
@stephendade Should be good now.

**Comment by stephendade** [comment]
Yep, help module works nicely now.

I have discovered another issue. Tab completion appears to be broken:
capture

(I tried tab completion for a variety of modules and settings, same error)

**Comment by SamuelDudley** [comment]
@stephendade Great! Thanks again for testing...  I'll have a look at that issue

**Comment by njoubert** [comment]
I am hugely in favor of this PR for one main reason. *Mac GUI support is possible with Python 3*. I have a demo running on my machine.

**Comment by SamuelDudley** [comment]
@njoubert when you say that you have a demo running on your machine, are you running the code from this PR or something else?

**Comment by SamuelDudley** [comment]
@stephendade when you have some time please try my latest commit. Thanks :+1:

**Comment by njoubert** [comment]
@SamuelDudley apologies, that was unclear. Not this code. In current MAVProxy, GUI elements doesn't work, since macOS does not allow forked processes to create GUI elements. Python 3's multiprocessing library allows using spawn rather than fork, and macOS does allow spawned processes to create GUI elements. I have a small test case that demonstrates this.

**Comment by SamuelDudley** [comment]
@njoubert In this PR I have forced the Python 3 multiprocessing library to use `spawn`, so you should be able to run it OK on macOS. https://github.com/ArduPilot/MAVProxy/pull/420/commits/e1ac4e3cfc4556404bd35fff01d972c17c9f6cde
I would be interested to know if it works for you / what other issues you run into...

**Comment by stephendade** [comment]
> @stephendade when you have some time please try my latest commit. Thanks 👍

Tab completion works under Windows now. All good :)

**Comment by SamuelDudley** [comment]
Great, thanks :)

**Comment by amilcarlucas** [comment]
How broken is python2 support at this moment ? is it possible to support both 2 and 3  ?

**Comment by SamuelDudley** [comment]
I think it's quite possible (especially if you run it headless / without the GUI modules). I'm planning on making a series of commits to clean up the prints first and then I'll look at supporting the non-gui elements in python 3. The GUI stuff is a little more complicated along with getting multiprocessing to play nice on all systems and between python 2/3...

**Comment by amilcarlucas** [comment]
Hi Samuel, can you rebase this ?

**Comment by SamuelDudley** [comment]
Sure thing, will do. I'm expecting quite a few conflicts

**Comment by SamuelDudley** [comment]
Im rebasing here https://github.com/SamuelDudley/MAVProxy/tree/wx_phoenix_wip_rebase
I need to test all of this again because there were a few merge conflicts. Once I'm happy I'll update this branch

**Comment by SamuelDudley** [comment]
@amilcarlucas rebased

File: MAVProxy/mavproxy.py [hunk 3]
```diff
@@ -9,17 +9,18 @@
 
 import sys, os, time, socket, signal
 import fnmatch, errno, threading
-import serial, Queue, select
+import serial, queue, select
 import traceback
 import select
 import shlex
 import platform
+import importlib
```
Comment (SamuelDudley) [inline MAVProxy/mavproxy.py]: this will work for python2.7+ and python3.x

**Comment by francisrod01** [comment]
I remember that I try to use with Python 3.6 version but without success.
https://github.com/ArduPilot/MAVProxy/issues/468

**Comment by tribbloid** [comment]
python2 support should be discarded, it almost reaches end-of-life.

**Comment by SamuelDudley** [comment]
In reality python2 will not go away overnight. I wish I had more time to take some of my changes in this branch and make usable patches to help the maintainers...
If anyone feels up to it it would help out the project quite a bit re python3 support!

**Comment by peterbarker** [comment]
On Sun, 25 Feb 2018, Samuel Dudley wrote:

> In reality python2 will not go away overnight. I wish I had more time to
> take some of my changes in this branch and make usable patches to help the
> maintainers...
> If anyone feels up to it it would help out the project quite a bit re
> python3 support!

People have suggested Qt instead of Phoenix.

Has anybody done an in-depth look at these options?

**Comment by francisrod01** [comment]
I'm expecting to use this version of the ArduPilot using MAVProxy with Python 3.6.x but until now this is not official, right?

**Comment by stephendade** [comment]
As an aside, wxPython 4 now appears to support Python 2.7 (https://pypi.python.org/pypi/wxPython/4.0.1), so no issue with GUI toolkits.

I'd really like to resurrect this PR and get it merged in - thoughts?

**Comment by SamuelDudley** [comment]
@stephendade I would love to see python 3.x support merged in but I personally don't have a much time to commit to this PR at the moment, sorry.
The biggest issue the 3.x support work will face IMO are the pickling errors. I have burned quite a few hours chasing them down without any real progress.

**Comment by SamuelDudley** [comment]
If you look at https://github.com/ArduPilot/MAVProxy/pull/420/files you can see that the number of changes is not that huge but it needs to be broken down into a set of PRs rather than this mess

**Comment by stephendade** [comment]
That's fine. Once I've finished my work with the cuav software, I'll see if I can split this up into manageable PR's and add some unit tests/CI to make the review easier.

**Comment by tridge** [comment]
i've got a new py3 PR here #560

**Comment by SamuelDudley** [comment]
@tridge are you happy for me to close this PR?

**Pull request closed without merging**